### PR TITLE
Multus images with Ubuntu and Debug option

### PR DESCRIPTION
**Pull request by phvalguima**
Adding the following to multus charm:
* move multus image to Ubuntu as base instead of centos
* debug option: set LogLevel to debug
* multus-debug image which ships multus with delve debugger

### CODE REVIEW

File: multus-debug/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ENV DEBIAN_FRONTEND "noninteractive"
+RUN ln -fs /usr/share/zoneinfo/UTC /etc/localtime
+
+ENV INSTALL_PKGS "git golang-1.13 python3 python3-yaml"
+
+RUN apt update && \
+    apt install -y $INSTALL_PKGS && \
+    ln -s /usr/lib/go-1.13/bin/go /usr/bin/go && \
+    ln -s /usr/bin/python3 /usr/bin/python && \
+    git clone https://github.com/k8snetworkplumbingwg/multus-cni /usr/src/multus-cni && \
```
Comment (Cynerva) [inline multus-debug/Dockerfile:14]: This git clone needs to pin to a specific tag of the repo, e.g.

```
git clone https://github.com/k8snetworkplumbingwg/multus-cni /usr/src/multus-cni --branch v3.7.1 --depth 1
```

File: multus-debug/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# This Dockerfile is used to build the image available on DockerHub
+FROM ubuntu:20.04 as build
+
+# Avoid asking for TZ info
+ENV DEBIAN_FRONTEND "noninteractive"
+RUN ln -fs /usr/share/zoneinfo/UTC /etc/localtime
+
+ENV INSTALL_PKGS "git golang-1.13 python3 python3-yaml"
```
Comment (Cynerva) [inline multus-debug/Dockerfile:8]: It looks like the upstream multus-cni is using golang 1.12, so we should probably do the same:

https://github.com/k8snetworkplumbingwg/multus-cni/blob/v3.7.1/go.mod#L3

Comment (Cynerva) [reply multus-debug/Dockerfile:8]: Nevermind. I tried building with golang 1.12 and encountered this:

```
#8 107.2 # k8s.io/apimachinery/pkg/util/errors
#8 107.2 /go/pkg/mod/k8s.io/apimachinery@v0.18.3/pkg/util/errors/errors.go:99:10: undefined: errors.Is
#8 107.2 note: module requires Go 1.13
```

1.13 it is. Consider adding a comment here to explain why we use 1.13.

File: multus-debug/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    apt install -y $INSTALL_PKGS && \
+    ln -s /usr/lib/go-1.13/bin/go /usr/bin/go && \
+    ln -s /usr/bin/python3 /usr/bin/python && \
+    git clone https://github.com/k8snetworkplumbingwg/multus-cni /usr/src/multus-cni && \
+    cd /usr/src/multus-cni && \
+    ./hack/build-go.sh
+
+# Building with go modules
+ENV GO111MODULE "on"
+RUN git clone https://github.com/go-delve/delve && \
```
Comment (Cynerva) [inline multus-debug/Dockerfile:20]: This git clone needs to be pinned to a specific tag, e.g.

```
git clone https://github.com/go-delve/delve --branch v1.6.0 --depth 1
```

File: multus/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+# Avoid asking for TZ info
+ENV DEBIAN_FRONTEND "noninteractive"
+RUN ln -fs /usr/share/zoneinfo/UTC /etc/localtime
+
+RUN apt update && \
+    apt install -y $INSTALL_PKGS && \
+    ln -s /usr/lib/go-1.13/bin/go /usr/bin/go && \
+    ln -s /usr/bin/python3 /usr/bin/python && \
+    git clone https://github.com/k8snetworkplumbingwg/multus-cni /usr/src/multus-cni && \
```
Comment (Cynerva) [inline multus/Dockerfile:14]: This git clone needs to pin to a specific tag, e.g.

```
git clone https://github.com/k8snetworkplumbingwg/multus-cni /usr/src/multus-cni --branch v3.7.1 --depth 1
```

File: config.yaml [hunk 5]
```diff
@@ -1,4 +1,12 @@
 options:
+  debug:
+    type: boolean
+    default: False
+    description: |
+      Set it to enable LogLevel=debug and reload debug image for containers.
+      Access kube-multus for debug with:
+      kubectl exec -n <model-name> -it <pod-name> -- bash
+      $ dlv attach --continue <PID>
```
Comment (Cynerva) [inline config.yaml:9]: I'm a bit confused about the usage of dlv here. What PID am I supposed to connect to? There is no persistent golang process:

```
$ kubectl exec -n k8s -it multus-jf8h4 -- ps aux
Defaulting container name to kube-multus.
Use 'kubectl describe pod/multus-jf8h4 -n k8s' to see all of the containers in this pod.
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.0  0.0   4100  3464 ?        Ss   20:00   0:00 /bin/bash /en
root          57  0.0  0.0   2508   592 ?        S    20:00   0:00 sleep infinit
root          64  0.0  0.0   5888  3068 pts/0    Rs+  20:06   0:00 ps aux
```

File: script/upload [hunk 6]
```diff
@@ -5,19 +5,25 @@ if ! charm whoami > /dev/null; then
...
 fi
-URL=$(charm push "$CHARM_BUILD_DIR/$CHARM.charm" "cs:~$NAMESPACE/$CHARM" | yq r - url)
-echo "Uploaded: $URL"
 
-MULTUS_IMAGE=$(yq r "metadata.yaml" "resources.multus-image.upstream-source")
-docker pull "$MULTUS_IMAGE"
-MULTUS_IMAGE_REV=$(charm attach cs:~"$NAMESPACE"/"$CHARM" --channel unpublished "multus-image=$MULTUS_IMAGE" | tail -n1 | sed -e 's/uploaded revision \([0-9]*\) of.*/\1/')
-echo "Attached: multus-image-$MULTUS_IMAGE_REV"
+
+URL=$(charm push "$CHARM_BUILD_DIR/$CHARM.charm" "cs:~$NAMESPACE/$CHARM" |  yq -r .url)
```
Comment (Cynerva) [inline script/upload:10]: This failed for me with yq version 4.6.1:

```
$ NAMESPACE=cynerva make upload
Verifying dependencies...
Error: unknown shorthand flag: 'r' in -r
Usage:
...
Use "yq [command] --help" for more information about a command.

make: *** [Makefile:16: upload] Error 1
```

I think it needs to be:

```
yq eval .url -
```

**Review by Cynerva — CHANGES_REQUESTED** [review]
This PR needs integration test coverage. For an example of how to do integration testing in charms, see how we do it in kubernetes-master:

* [test_kubernetes_master_integration.py](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/master/tests/integration/test_kubernetes_master_integration.py)
* [tox.ini](https://github.com/charmed-kubernetes/charm-kubernetes-master/blob/91e7ae2543db2629504ab6efd9ecac8687396fa0/tox.ini#L21-L28)

**Comment by kwmonroe** [comment]
This PR now conflicts with the main repo and appears to have gone stale wrt requested changes.  I'm closing for now; feel free to reopen if there is still work in progress here.

**Pull request closed without merging**